### PR TITLE
Enhancement/7349-dashboard-sharing-kmw

### PR DESCRIPTION
**Pull request by jimmymadon**
## Summary

Addresses issue:

- #7349

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Key_Metrics/Is_Key_Metrics_Setup_Complete.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Is_Key_Metrics_Setup_Complete extends Setting {
+
+	/**
+	 * The option_name for this setting.
+	 */
+	const OPTION = 'googlesitekit_is_key_metrics_setup_complete';
```
Comment (aaemnnosttv) [inline includes/Core/Key_Metrics/Is_Key_Metrics_Setup_Complete.php:27]: Seeing this now, it feels a bit wasteful to use a new option just for a boolean and perhaps we could use a new "flags" option for this kind of thing similar to how we have for feature tours and dismissed items.

I think refactoring this here is out of scope but seems like a good future enhancement.

File: includes/Core/User_Input/User_Input.php [hunk 2]
```diff
@@ -97,7 +98,8 @@ public function __construct(
 		$this->user_specific_answers = new User_Specific_Answers( $this->user_options );
 		$this->rest_controller       = new REST_User_Input_Controller(
 			$this,
-			$survey_queue ?: new Survey_Queue( $this->user_options )
+			$survey_queue ?: new Survey_Queue( $this->user_options ),
+			new Is_Key_Metrics_Setup_Complete( $options ?: new Options( $context ) )
```
Comment (aaemnnosttv) [inline includes/Core/User_Input/User_Input.php:102]: Thinking about this again, let's not couple user input with key metrics directly here.

Instead, I think we can listen for the site-specific answers being saved/updated (`googlesitekit_user_input_settings`) and use that to conditionally set the flag in `Key_Metrics`. That should be a relatively small change to make and keep the logic for this under the KM realm of the codebase.

**Review by aaemnnosttv — COMMENTED** [review]
Thanks @jimmymadon – a few suggestions for you but this is looking good 👍

File: includes/Core/User_Input/User_Input.php (hunk 2 above)
Comment (aaemnnosttv) [reply includes/Core/User_Input/User_Input.php]: Leaving as defined in IB for now.

**Review by aaemnnosttv — APPROVED** [review]
This should be good to go now and works well in my testing 👍

**Review by techanvil — APPROVED** [review]
LGTM, nice one @jimmymadon & @aaemnnosttv!

File: includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php [hunk 3]
```diff
@@ -107,11 +120,19 @@ protected function get_rest_routes() {
...
 						$settings = $data['settings'];
 
+						$num_widgets = count( $settings['widgetSlugs'] );
+						if ( ! $num_widgets ) {
+							return new WP_Error(
+								'rest_invalid_param',
+								__( 'Selected metrics cannot be empty.', 'google-site-kit' ),
+								array( 'status' => 400 )
+							);
+						}
```
[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:130]: Hmm, while bug bashing I've realised this causes a problem when toggling the `isWidgetHidden` from the Admin page when only the questionnaire has been answered and no key metrics are explicitly set (i.e. this array is empty).

I've raised this as an issue on the bug bash board.

[the remark this answers is not in the document]
Comment (aaemnnosttv) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:130]: Ah that makes sense. We should probably store whether KMW is using answer-based or user-selections, then we could only validate this in the case where it's using hand-picked metrics 🤔  There's other ways around this but the idea here is that we're only marking the setup completed here when the user is completing by picking their own metrics.

[the remark this answers is not in the document]
Comment (techanvil) [reply includes/Core/Key_Metrics/REST_Key_Metrics_Controller.php:130]: Yup, sounds good. I've created a new GH [issue](https://github.com/google/site-kit-wp/issues/7441) for this where we can flesh out the specific approach in the IB.

**Pull request merged**